File: KOI/2017/14864.py
# https://www.acmicpc.net/problem/14864
import sys

flag = True
n, m = map(int, sys.stdin.readline().split())

lists = [i for i in range(n + 1)]

for i in range(0, m):
    a, b = map(int, sys.stdin.readline().split())
    lists[a], lists[b] = lists[a] + 1, lists[b] - 1


check = [False] * (m + 1)

for i in range(1, n + 1):
    j = lists[i]
    if check[j] == False:
        check[j] = True
        continue
    if check[j] == True:
        flag = False
        break

if flag:
    for i in range(1, n + 1):
        print(lists[i], end=" ")
    print()
else:
    print(-1)

File: 2019/KOI/2017/14864.cpp
//https://www.acmicpc.net/problem/14864
#include <bits/stdc++.h>
using namespace std;

int student[100001];
bool check[100001];
int n, m;

void init();
int main(void)
{
    bool flag = false;
    cin >> n >> m;

    init();

    while (m--)
    {
        int a, b;
        cin >> a >> b;
        student[a] += 1, student[b] -= 1;
    }

    for (int i = 1; i <= n; i++)
    {
        int j = student[i];

        if (j <= 0 || j > n)
        {
            flag = true;
            break;
        }

        else if (!check[j])
        {
            check[j] = true;
            continue;
        }
        else if (check[j])
        {
            flag = true;
            break;
        }
    }

    if (!flag)
    {
        for (int i = 1; i <= n; i++)
            cout << student[i] << " ";
        cout << '\n';
    }

    else
        cout << "-1\n";

    return 0;
}

void init()
{
    for (int i = 1; i <= n; i++)
        student[i] = i;
}
